transport/pool.go

```diff
@@ -0,0 +1,252 @@
+
+	"go.uber.org/atomic"
+)
+
+// Node Connection Pool - One per node, Node struct holds pointer to it
+// This struct holds connection to shards along with helper variables.
+type ConnPool struct {
+	conns     []atomic.Value
+	alertChan chan uint16
+
+	nrShards  uint16
+	msbIgnore uint8
+
+	idx atomic.Uint64
+}
+
+// Atomicly stores pointer to a connection.
+func (c *ConnPool) storeConn(idx uint16, conn *Conn) {
+	c.conns[idx].Store(conn)
+}
+
+// Atomicly load pointer to a connection.
+func (c *ConnPool) loadConn(idx uint16) *Conn {
+	conn := c.conns[idx].Load()
+	if conn == nil {
+		return nil
+	}
+	return conn.(*Conn)
+}
+
+// RandConn returns a random connection but has no protection
+// of it being nil, that's why NextConn() might be a better choice for now.
+func (c *ConnPool) RandConn() *Conn {
+	idx := rand.Uint32() % uint32(c.nrShards)
+	return c.loadConn(uint16(idx))
+}
+
+// NextConn returns nearest (from idx % nrShards) not nil connection
+// if it iterates over whole slice and doesn't find any (extremely unlikely) it returns nil.
+func (c *ConnPool) NextConn() *Conn {
+	idx := c.idx.Add(1)
+	for i := uint16(0); i < c.nrShards; i++ {
+		if conn := c.loadConn(uint16((idx + uint64(i)) % uint64(c.nrShards))); conn != nil {
+			return conn
+		}
+	}
+	return nil
+}
+
+// Calculates shard index in ConnPool.conns slice based on given token
+// Assumes it's a MurmurHash3 token.
+func (c *ConnPool) shardOf(token Token) int {
+	shards := uint64(c.nrShards)
+	z := uint64(token.value+math.MinInt64) << c.msbIgnore
+	lo := z & 0xffffffff
+	hi := (z >> 32) & 0xffffffff
+	mul1 := lo * shards
+	mul2 := hi * shards
+	sum := (mul1 >> 32) + mul2
+	return int(sum >> 32)
+}
+
+// Given shard token returns connection under this token if
+// it's found and not nil, otherwise returns random connection.
+func (c *ConnPool) GetConn(token Token) *Conn {
+	idx := c.shardOf(token)
+	if conn := c.loadConn(uint16(idx)); conn != nil {
+		// We have this shard's connection
+		// so let's give it to the caller.
+		return conn
+	}
+	return c.RandConn()
+}
+
+// Helper in closing connection, actual closing happens
+// in PoolRefiller because of multithreading issues.
+// When wanted connection is found it passes its index to PoolRefiller channel.
+func (c *ConnPool) Close(conn *Conn) {
+	for i, v := range c.conns {
+		if v.Load() == conn {
+			c.alertChan <- uint16(i)
+		}
+	}
+}
+
+// Handles errors and occasionally fills ConnPool.
+type PoolRefiller struct {
+	address    string
+	connConfig ConnConfig
+	connPool   ConnPool
+
+	activeShards uint16
+
+	connOptions      frame.ScyllaSupported
+	closeSinceRefill bool
+	currentDelay     time.Duration
+}
+
+// Fill happens in changing time intervals, these are its helper constants.
+const (
+	minFillBackoff        = 50 * time.Millisecond
+	maxFillBackoff        = 10 * time.Second
+	fillBackoffMultiplier = 2
+)
+
+// PoolRefiller endless function, for now it never shuts down.
+func (p *PoolRefiller) loop() {
```

Comment:
This is a quite useless comment `PoolRefiller endless function, for now it never shuts down.`

BUT

every go loop function must be able to exit - think we close the session and.